cmd/kubeadm/app/cmd/phases/bootstraptoken.go

```diff
@@ -107,14 +264,96 @@ func NewSubCmdNodeBootstrapTokenPostCSRs(kubeConfigFile *string) *cobra.Command
+	)
+	flagSet.DurationVar(
+		&cfg.TokenTTL.Duration, "ttl", kubeadmconstants.DefaultTokenDuration,
+		"The duration before the token is automatically deleted (e.g. 1s, 2m, 3h). If set to '0', the token will never expire",
+	)
+	flagSet.StringSliceVar(
+		usages, "usages", kubeadmconstants.DefaultTokenUsages,
+		fmt.Sprintf("Describes the ways in which this token can be used. You can pass --usages multiple times or provide a comma separated list of options. Valid options: [%s]", strings.Join(kubeadmconstants.DefaultTokenUsages, ",")),
+	)
+	flagSet.StringSliceVar(
+		extraGroups, "groups", []string{kubeadmconstants.NodeBootstrapTokenAuthGroup},
+		fmt.Sprintf("Extra groups that this token will authenticate as when used for authentication. Must match %q", bootstrapapi.BootstrapGroupPattern),
+	)
+	flagSet.StringVar(
+		description, "description", "The default bootstrap token generated by 'kubeadm init'.",
+		"A human friendly description of how this token is used.",
+	)
+	flagSet.BoolVar(
+		skipTokenPrint, "skip-token-print", *skipTokenPrint,
+		"Skip printing of the bootstrap token",
+	)
+}
+
+func createBootstrapToken(client clientset.Interface, cfgPath string, cfg *kubeadmapiext.MasterConfiguration, description string, usages, extraGroups []string, skipTokenPrint bool) error {
+
+	// adding groups only makes sense for authentication
+	usagesSet := sets.NewString(usages...)
+	usageAuthentication := strings.TrimPrefix(bootstrapapi.BootstrapTokenUsageAuthentication, bootstrapapi.BootstrapTokenUsagePrefix)
+	if len(extraGroups) > 0 && !usagesSet.Has(usageAuthentication) {
+		return fmt.Errorf("--groups cannot be specified unless --usages includes %q", usageAuthentication)
+	}
+
+	// validate any extra group names
+	for _, group := range extraGroups {
+		if err := bootstrapapi.ValidateBootstrapGroupName(group); err != nil {
+			return err
+		}
+	}
+
+	// This call returns the ready-to-use configuration based on the configuration file that might or might not exist and the default cfg populated by flags
+	internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
+	kubeadmutil.CheckErr(err)
+
+	// Load the CA certificate from so we can pin its public key
+	caCert, err := pkiutil.TryLoadCertFromDisk(internalcfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName)
+	if err != nil {
+		return fmt.Errorf("error loading ca cert from disk: %v", err)
+	}
+
+	// Creates or updates the token
+	if err := node.UpdateOrCreateToken(client, internalcfg.Token, false, internalcfg.TokenTTL.Duration, usages, extraGroups, description); err != nil {
+		return err
+	}
+
+	fmt.Println("[bootstraptoken] Bootstrap token Created")
+	if skipTokenPrint {
+		internalcfg.Token = "{token}"
+	}
+	fmt.Println("[bootstraptoken] You can now join any number of machines by running:")
+	fmt.Printf("[bootstraptoken]   kubeadm join {master} --token %s --discovery-token-ca-cert-hash %s \n", internalcfg.Token, pubkeypin.Hash(caCert))
```

Comment:
Seems like https://github.com/kubernetes/kubeadm/issues/519 would be fixed here? cc @luxas

Reply:
nope; this is for when you create a **new** token, but the issue is about getting it back with the original token

Reply:
What's the final decision of that thread? Maybe this idea suggested by @mattmoyer in https://github.com/kubernetes/kubeadm/issues/519#issuecomment-342832458 ?
```
A new command kubeadm token get <token id> that gets the full token given the ID, with a --print-join-command flag to optionally print the full join command instead of just the token.
```

Reply:
If we want @mattmoyer 's second idea, I think it will be the same logic here. Am I missing something (again)? :D
```
Add a --print-join-command flag for kubeadm token create that prints out the whole join command just like kubeadm init does. This would hopefully make it easy to automate around by running something like kubeadm token create --print-join-command >> worker_init.sh during provisioning.
```